File: packages/adapter-anthropic-agent/src/adapter.ts
/**
 * Anthropic Agent SDK Adapter for Open Harness
 *
 * Wraps the Anthropic SDK to provide Open Harness compatibility.
 */

import Anthropic from "@anthropic-ai/sdk";
import type {
  HarnessAdapter,
  AdapterCapabilities,
  ExecuteRequest,
  ExecutionEvent,
  AdapterExecutionResult,
  CapabilityManifest,
  Tool,
  createHarnessId,
} from "@openharness/client";

import type {
  AnthropicAgentConfig,
  AdapterTool,
  ExecutionOptions,
  ToolResult,
  Conversation,
  ConversationMessage,
  PendingToolCall,
} from "./types";

// Re-export for convenience
export type { HarnessAdapter } from "@openharness/client";

/**
 * Default model to use
 */
const DEFAULT_MODEL = "claude-sonnet-4-20250514";

/**
 * Default max tokens
 */
const DEFAULT_MAX_TOKENS = 4096;

/**
 * Anthropic Agent SDK Adapter
 *
 * Provides Open Harness compatibility for the Anthropic SDK, enabling
 * tool use, streaming, and multi-turn conversations.
 *
 * @example
 * ```ts
 * import { AnthropicAgentAdapter } from "@openharness/adapter-anthropic-agent";
 *
 * const adapter = new AnthropicAgentAdapter({
 *   apiKey: process.env.ANTHROPIC_API_KEY,
 *   model: "claude-sonnet-4-20250514",
 * });
 *
 * // Simple execution
 * const result = await adapter.execute({
 *   message: "What is 2 + 2?",
 * });
 * console.log(result.output);
 *
 * // Streaming execution
 * for await (const event of adapter.executeStream({
 *   message: "Write a haiku about coding",
 * })) {
 *   if (event.type === "text") {
 *     process.stdout.write(event.content);
 *   }
 * }
 * ```
 */
export class AnthropicAgentAdapter implements Partial<HarnessAdapter> {
  readonly id: ReturnType<typeof createHarnessId>;
  readonly name = "Anthropic Agent SDK";
  readonly version = "0.1.0";

  readonly capabilities: AdapterCapabilities = {
    agents: false, // Agent management is not built-in to the SDK
    skills: false, // Skills require external management
    execution: true,
    streaming: true,
    sessions: true, // We can manage conversations
    memory: false, // Memory requires external storage
    subagents: false, // Subagents require orchestration layer
    mcp: false, // MCP requires @anthropic-ai/mcp package
    files: false, // File operations require external tools
    hooks: false, // Hooks require orchestration layer
    planning: false, // Planning requires orchestration layer
    websocket: false,
    multipart: false,
    binaryDownload: false,
  };

  private client: Anthropic;
  private config: Required<
    Pick<AnthropicAgentConfig, "model" | "maxTokens">
  > &
    AnthropicAgentConfig;
  private tools: Map<string, AdapterTool> = new Map();
  private conversations: Map<string, Conversation> = new Map();

  constructor(config: AnthropicAgentConfig = {}) {
    // Use provided client or create new one
    this.client =
      config.client ??
      new Anthropic({
        apiKey: config.apiKey,
        baseURL: config.baseUrl,
      });

    this.config = {
      ...config,
      model: config.model ?? DEFAULT_MODEL,
      maxTokens: config.maxTokens ?? DEFAULT_MAX_TOKENS,
    };

    // Create a branded harness ID
    this.id = "anthropic-agent" as ReturnType<typeof createHarnessId>;
  }

  /**
   * Get the capability manifest for this adapter
   */
  async getCapabilityManifest(): Promise<CapabilityManifest> {
    return {
      harness_id: this.id,
      version: this.version,
      capabilities: [
        { id: "execution.run", supported: true },
        { id: "execution.stream", supported: true },
        { id: "tools.list", supported: true },
        { id: "tools.invoke", supported: true },
        { id: "sessions.create", supported: true },
        { id: "sessions.sendMessage", supported: true },
        { id: "sessions.sendMessageStream", supported: true },
      ],
    };
  }

  // ===========================================================================
  // Tool Management
  // ===========================================================================

  /**
   * Register a tool with the adapter
   */
  registerTool(tool: AdapterTool): void {
    this.tools.set(tool.name, tool);
  }

  /**
   * Unregister a tool
   */
  unregisterTool(name: string): boolean {
    return this.tools.delete(name);
  }

  /**
   * List all registered tools
   */
  async listTools(): Promise<Tool[]> {
    return Array.from(this.tools.values()).map((tool) => ({
      id: tool.name,
      name: tool.name,
      description: tool.description,
      source: "custom" as const,
      input_schema: tool.inputSchema as object,
    }));
  }

  /**
   * Invoke a tool directly
   */
  async invokeTool(
    toolId: string,
    input: Record<string, unknown>
  ): Promise<ToolResult> {
    const tool = this.tools.get(toolId);
    if (!tool) {
      return {
        success: false,
        error: `Tool not found: ${toolId}`,
      };
    }

    try {
      return await tool.handler(input);
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  // ===========================================================================
  // Execution
  // ===========================================================================

  /**
   * Execute a request and return the complete result
   */
  async execute(
    request: ExecuteRequest,
    options?: ExecutionOptions
  ): Promise<AdapterExecutionResult> {
    const startTime = Date.now();
    let totalInputTokens = 0;
    let totalOutputTokens = 0;
    const toolCalls: AdapterExecutionResult["toolCalls"] = [];

    // Build messages
    const messages: Anthropic.MessageParam[] = [
      { role: "user", content: request.message },
    ];

    // Build system prompt
    const systemPrompt = this.buildSystemPrompt(
      request.system_prompt,
      options?.systemPrompt
    );

    // Get tools for this execution
    const tools = this.getToolsForExecution(options?.tools);

    // Agentic loop - continue until no more tool calls
    let iteration = 0;
    const maxIterations = options?.maxIterations ?? 10;
    let finalContent = "";

    while (iteration < maxIterations) {
      iteration++;

      const response = await this.client.messages.create({
        model: options?.model ?? this.config.model,
        max_tokens: options?.maxTokens ?? this.config.maxTokens,
        temperature: options?.temperature,
        system: systemPrompt,
        messages,
        tools: tools.length > 0 ? tools : undefined,
      });

      totalInputTokens += response.usage.input_tokens;
      totalOutputTokens += response.usage.output_tokens;

      // Process response content
      const assistantContent: Anthropic.ContentBlock[] = [];
      const pendingToolCalls: PendingToolCall[] = [];

      for (const block of response.content) {
        assistantContent.push(block);

        if (block.type === "text") {
          finalContent += block.text;
        } else if (block.type === "tool_use") {
          pendingToolCalls.push({
            id: block.id,
            name: block.name,
            input: block.input as Record<string, unknown>,
          });
        }
      }

      // Add assistant message
      messages.push({ role: "assistant", content: assistantContent });

      // If no tool calls, we're done
      if (pendingToolCalls.length === 0 || response.stop_reason === "end_turn") {
        break;
      }

      // Execute tool calls
      const toolResults: Anthropic.ToolResultBlockParam[] = [];

      for (const toolCall of pendingToolCalls) {
        const result = await this.invokeTool(toolCall.name, toolCall.input);

        toolCalls.push({
          id: toolCall.id,
          name: toolCall.name,
          input: toolCall.input,
          output: result.output as object | undefined,
          error: result.error,
        });

        toolResults.push({
          type: "tool_result",
          tool_use_id: toolCall.id,
          content: result.success
            ? JSON.stringify(result.output)
            : `Error: ${result.error}`,
          is_error: !result.success,
        });
      }

      // Add tool results as user message
      messages.push({ role: "user", content: toolResults });
    }

    return {
      output: finalContent,
      toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
      usage: {
        inputTokens: totalInputTokens,
        outputTokens: totalOutputTokens,
        totalTokens: totalInputTokens + totalOutputTokens,
        durationMs: Date.now() - startTime,
      },
    };
  }

  /**
   * Execute a request with streaming
   */
  async *executeStream(
    request: ExecuteRequest,
    options?: ExecutionOptions
  ): AsyncGenerator<ExecutionEvent> {
    const startTime = Date.now();
    let totalInputTokens = 0;
    let totalOutputTokens = 0;

    // Build messages
    const messages: Anthropic.MessageParam[] = [
      { role: "user", content: request.message },
    ];

    // Build system prompt
    const systemPrompt = this.buildSystemPrompt(
      request.system_prompt,
      options?.systemPrompt
    );

    // Get tools for this execution
    const tools = this.getToolsForExecution(options?.tools);

    // Agentic loop
    let iteration = 0;
    const maxIterations = options?.maxIterations ?? 10;
    let totalSteps = 1; // We don't know total steps upfront

    while (iteration < maxIterations) {
      iteration++;

      yield {
        type: "progress",
        percentage: Math.min((iteration / maxIterations) * 100, 99),
        step: `Iteration ${iteration}`,
        step_number: iteration,
        total_steps: totalSteps,
      };

      // Create streaming request
      const stream = this.client.messages.stream({
        model: options?.model ?? this.config.model,
        max_tokens: options?.maxTokens ?? this.config.maxTokens,
        temperature: options?.temperature,
        system: systemPrompt,
        messages,
        tools: tools.length > 0 ? tools : undefined,
      });

      const assistantContent: Anthropic.ContentBlock[] = [];
      const pendingToolCalls: PendingToolCall[] = [];
      let currentToolCall: Partial<PendingToolCall> | null = null;

      // Process stream events
      for await (const event of stream) {
        if (options?.abortSignal?.aborted) {
          yield {
            type: "error",
            code: "cancelled",
            message: "Execution was cancelled",
            recoverable: false,
          };
          return;
        }

        switch (event.type) {
          case "content_block_start":
            if (event.content_block.type === "text") {
              // Text block starting
            } else if (event.content_block.type === "tool_use") {
              currentToolCall = {
                id: event.content_block.id,
                name: event.content_block.name,
                input: {},
              };
              yield {
                type: "tool_call_start",
                id: event.content_block.id,
                name: event.content_block.name,
                input: {},
              };
            }
            break;

          case "content_block_delta":
            if (event.delta.type === "text_delta") {
              yield {
                type: "text",
                content: event.delta.text,
              };
            } else if (event.delta.type === "input_json_delta") {
              if (currentToolCall) {
                yield {
                  type: "tool_call_delta",
                  id: currentToolCall.id!,
                  input_delta: { partial_json: event.delta.partial_json },
                };
              }
            }
            break;

          case "content_block_stop":
            if (currentToolCall?.id) {
              yield {
                type: "tool_call_end",
                id: currentToolCall.id,
              };
            }
            break;

          case "message_delta":
            // Message is ending
            break;
        }
      }

      // Get final message
      const finalMessage = await stream.finalMessage();
      totalInputTokens += finalMessage.usage.input_tokens;
      totalOutputTokens += finalMessage.usage.output_tokens;

      // Process final content for tool calls
      for (const block of finalMessage.content) {
        assistantContent.push(block);
        if (block.type === "tool_use") {
          pendingToolCalls.push({
            id: block.id,
            name: block.name,
            input: block.input as Record<string, unknown>,
          });
        }
      }

      // Add assistant message to history
      messages.push({ role: "assistant", content: assistantContent });

      // If no tool calls, we're done
      if (
        pendingToolCalls.length === 0 ||
        finalMessage.stop_reason === "end_turn"
      ) {
        break;
      }

      // Execute tool calls
      const toolResults: Anthropic.ToolResultBlockParam[] = [];

      for (const toolCall of pendingToolCalls) {
        const result = await this.invokeTool(toolCall.name, toolCall.input);

        yield {
          type: "tool_result",
          id: toolCall.id,
          success: result.success,
          output: result.output as object | undefined,
        };

        toolResults.push({
          type: "tool_result",
          tool_use_id: toolCall.id,
          content: result.success
            ? JSON.stringify(result.output)
            : `Error: ${result.error}`,
          is_error: !result.success,
        });
      }

      // Add tool results
      messages.push({ role: "user", content: toolResults });
      totalSteps = iteration + 1;
    }

    // Done
    yield {
      type: "done",
      usage: {
        input_tokens: totalInputTokens,
        output_tokens: totalOutputTokens,
        total_tokens: totalInputTokens + totalOutputTokens,
        duration_ms: Date.now() - startTime,
      },
    };
  }

  // ===========================================================================
  // Conversation Management
  // ===========================================================================

  /**
   * Create a new conversation
   */
  createConversation(options?: {
    systemPrompt?: string;
    tools?: AdapterTool[];
  }): Conversation {
    const id = crypto.randomUUID();
    const conversation: Conversation = {
      id,
      messages: [],
      systemPrompt: options?.systemPrompt,
      tools: options?.tools,
      createdAt: new Date(),
      updatedAt: new Date(),
    };
    this.conversations.set(id, conversation);
    return conversation;
  }

  /**
   * Get a conversation by ID
   */
  getConversation(id: string): Conversation | undefined {
    return this.conversations.get(id);
  }

  /**
   * Delete a conversation
   */
  deleteConversation(id: string): boolean {
    return this.conversations.delete(id);
  }

  /**
   * Send a message in a conversation
   */
  async sendMessage(
    conversationId: string,
    content: string,
    options?: ExecutionOptions
  ): Promise<AdapterExecutionResult> {
    const conversation = this.conversations.get(conversationId);
    if (!conversation) {
      throw new Error(`Conversation not found: ${conversationId}`);
    }

    // Add user message
    conversation.messages.push({ role: "user", content });

    // Execute with conversation history
    const result = await this.executeWithHistory(
      conversation.messages,
      conversation.systemPrompt,
      conversation.tools,
      options
    );

    // Add assistant response
    conversation.messages.push({ role: "assistant", content: result.output });
    conversation.updatedAt = new Date();

    return result;
  }

  /**
   * Send a message with streaming in a conversation
   */
  async *sendMessageStream(
    conversationId: string,
    content: string,
    options?: ExecutionOptions
  ): AsyncGenerator<ExecutionEvent> {
    const conversation = this.conversations.get(conversationId);
    if (!conversation) {
      yield {
        type: "error",
        code: "conversation_not_found",
        message: `Conversation not found: ${conversationId}`,
        recoverable: false,
      };
      return;
    }

    // Add user message
    conversation.messages.push({ role: "user", content });

    // Stream with conversation history
    let fullResponse = "";
    for await (const event of this.executeStreamWithHistory(
      conversation.messages,
      conversation.systemPrompt,
      conversation.tools,
      options
    )) {
      if (event.type === "text") {
        fullResponse += event.content;
      }
      yield event;
    }

    // Add assistant response
    conversation.messages.push({ role: "assistant", content: fullResponse });
    conversation.updatedAt = new Date();
  }

  // ===========================================================================
  // Private Helpers
  // ===========================================================================

  private buildSystemPrompt(
    requestPrompt?: string,
    optionsPrompt?: string
  ): string | undefined {
    const parts: string[] = [];

    if (this.config.systemPrompt) {
      parts.push(this.config.systemPrompt);
    }
    if (requestPrompt) {
      parts.push(requestPrompt);
    }
    if (optionsPrompt) {
      parts.push(optionsPrompt);
    }

    return parts.length > 0 ? parts.join("\n\n") : undefined;
  }

  private getToolsForExecution(
    additionalTools?: AdapterTool[]
  ): Anthropic.Tool[] {
    const allTools = new Map(this.tools);

    // Add any additional tools for this execution
    if (additionalTools) {
      for (const tool of additionalTools) {
        allTools.set(tool.name, tool);
      }
    }

    return Array.from(allTools.values()).map((tool) => ({
      name: tool.name,
      description: tool.description,
      input_schema: tool.inputSchema,
    }));
  }

  private async executeWithHistory(
    history: ConversationMessage[],
    systemPrompt?: string,
    conversationTools?: AdapterTool[],
    options?: ExecutionOptions
  ): Promise<AdapterExecutionResult> {
    const startTime = Date.now();
    let totalInputTokens = 0;
    let totalOutputTokens = 0;
    const toolCalls: AdapterExecutionResult["toolCalls"] = [];

    // Convert history to Anthropic format
    const messages: Anthropic.MessageParam[] = history.map((msg) => ({
      role: msg.role,
      content: msg.content,
    }));

    // Build system prompt
    const fullSystemPrompt = this.buildSystemPrompt(
      systemPrompt,
      options?.systemPrompt
    );

    // Get tools
    const tools = this.getToolsForExecution(
      conversationTools ?? options?.tools
    );

    // Agentic loop
    let iteration = 0;
    const maxIterations = options?.maxIterations ?? 10;
    let finalContent = "";

    while (iteration < maxIterations) {
      iteration++;

      const response = await this.client.messages.create({
        model: options?.model ?? this.config.model,
        max_tokens: options?.maxTokens ?? this.config.maxTokens,
        temperature: options?.temperature,
        system: fullSystemPrompt,
        messages,
        tools: tools.length > 0 ? tools : undefined,
      });

      totalInputTokens += response.usage.input_tokens;
      totalOutputTokens += response.usage.output_tokens;

      const assistantContent: Anthropic.ContentBlock[] = [];
      const pendingToolCalls: PendingToolCall[] = [];

      for (const block of response.content) {
        assistantContent.push(block);
        if (block.type === "text") {
          finalContent += block.text;
        } else if (block.type === "tool_use") {
          pendingToolCalls.push({
            id: block.id,
            name: block.name,
            input: block.input as Record<string, unknown>,
          });
        }
      }

      messages.push({ role: "assistant", content: assistantContent });

      if (pendingToolCalls.length === 0 || response.stop_reason === "end_turn") {
        break;
      }

      const toolResults: Anthropic.ToolResultBlockParam[] = [];

      for (const toolCall of pendingToolCalls) {
        const result = await this.invokeTool(toolCall.name, toolCall.input);

        toolCalls.push({
          id: toolCall.id,
          name: toolCall.name,
          input: toolCall.input,
          output: result.output as object | undefined,
          error: result.error,
        });

        toolResults.push({
          type: "tool_result",
          tool_use_id: toolCall.id,
          content: result.success
            ? JSON.stringify(result.output)
            : `Error: ${result.error}`,
          is_error: !result.success,
        });
      }

      messages.push({ role: "user", content: toolResults });
    }

    return {
      output: finalContent,
      toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
      usage: {
        inputTokens: totalInputTokens,
        outputTokens: totalOutputTokens,
        totalTokens: totalInputTokens + totalOutputTokens,
        durationMs: Date.now() - startTime,
      },
    };
  }

  private async *executeStreamWithHistory(
    history: ConversationMessage[],
    systemPrompt?: string,
    conversationTools?: AdapterTool[],
    options?: ExecutionOptions
  ): AsyncGenerator<ExecutionEvent> {
    // Build messages from history
    const messages: Anthropic.MessageParam[] = history.map((msg) => ({
      role: msg.role,
      content: msg.content,
    }));

    const startTime = Date.now();
    let totalInputTokens = 0;
    let totalOutputTokens = 0;

    const fullSystemPrompt = this.buildSystemPrompt(
      systemPrompt,
      options?.systemPrompt
    );
    const tools = this.getToolsForExecution(
      conversationTools ?? options?.tools
    );

    let iteration = 0;
    const maxIterations = options?.maxIterations ?? 10;

    while (iteration < maxIterations) {
      iteration++;

      const stream = this.client.messages.stream({
        model: options?.model ?? this.config.model,
        max_tokens: options?.maxTokens ?? this.config.maxTokens,
        temperature: options?.temperature,
        system: fullSystemPrompt,
        messages,
        tools: tools.length > 0 ? tools : undefined,
      });

      const assistantContent: Anthropic.ContentBlock[] = [];
      const pendingToolCalls: PendingToolCall[] = [];
      let currentToolCall: Partial<PendingToolCall> | null = null;

      for await (const event of stream) {
        switch (event.type) {
          case "content_block_start":
            if (event.content_block.type === "tool_use") {
              currentToolCall = {
                id: event.content_block.id,
                name: event.content_block.name,
              };
              yield {
                type: "tool_call_start",
                id: event.content_block.id,
                name: event.content_block.name,
                input: {},
              };
            }
            break;

          case "content_block_delta":
            if (event.delta.type === "text_delta") {
              yield { type: "text", content: event.delta.text };
            }
            break;

          case "content_block_stop":
            if (currentToolCall?.id) {
              yield { type: "tool_call_end", id: currentToolCall.id };
              currentToolCall = null;
            }
            break;
        }
      }

      const finalMessage = await stream.finalMessage();
      totalInputTokens += finalMessage.usage.input_tokens;
      totalOutputTokens += finalMessage.usage.output_tokens;

      for (const block of finalMessage.content) {
        assistantContent.push(block);
        if (block.type === "tool_use") {
          pendingToolCalls.push({
            id: block.id,
            name: block.name,
            input: block.input as Record<string, unknown>,
          });
        }
      }

      messages.push({ role: "assistant", content: assistantContent });

      if (
        pendingToolCalls.length === 0 ||
        finalMessage.stop_reason === "end_turn"
      ) {
        break;
      }

      const toolResults: Anthropic.ToolResultBlockParam[] = [];

      for (const toolCall of pendingToolCalls) {
        const result = await this.invokeTool(toolCall.name, toolCall.input);

        yield {
          type: "tool_result",
          id: toolCall.id,
          success: result.success,
          output: result.output as object | undefined,
        };

        toolResults.push({
          type: "tool_result",
          tool_use_id: toolCall.id,
          content: result.success
            ? JSON.stringify(result.output)
            : `Error: ${result.error}`,
          is_error: !result.success,
        });
      }

      messages.push({ role: "user", content: toolResults });
    }

    yield {
      type: "done",
      usage: {
        input_tokens: totalInputTokens,
        output_tokens: totalOutputTokens,
        total_tokens: totalInputTokens + totalOutputTokens,
        duration_ms: Date.now() - startTime,
      },
    };
  }
}

File: packages/adapter-anthropic-agent-python/src/openharness_anthropic_agent/adapter.py
"""
Anthropic Agent adapter for Open Harness.

Provides direct access to Claude via the Anthropic Python SDK with
support for tool use, streaming, and extended thinking.
"""

from __future__ import annotations

import os
from typing import Any, AsyncIterator

import anthropic

from openharness.adapter import (
    AdapterCapabilities,
    AdapterExecutionResult,
    HarnessAdapter,
    ToolDefinition,
)
from openharness.types import (
    CapabilityInfo,
    CapabilityManifest,
    ExecuteRequest,
    Tool,
    UsageStats,
)
from openharness.types.events import (
    DoneEvent,
    ErrorEvent,
    ExecutionEvent,
    TextEvent,
    ThinkingEvent,
    ToolCallEndEvent,
    ToolCallStartEvent,
    ToolResultEvent,
)


class AnthropicAgentAdapter(HarnessAdapter):
    """
    Open Harness adapter for Anthropic Claude API.

    Provides direct access to Claude with:
    - Tool use with automatic execution loop
    - Streaming responses
    - Extended thinking support
    - Conversation management

    Example:
        ```python
        from openharness_anthropic_agent import AnthropicAgentAdapter
        from openharness import ExecuteRequest

        adapter = AnthropicAgentAdapter(
            api_key=os.environ["ANTHROPIC_API_KEY"],
            model="claude-sonnet-4-20250514",
        )

        # Streaming
        async for event in adapter.execute_stream(
            ExecuteRequest(message="Hello!")
        ):
            if event.type == "text":
                print(event.content, end="")

        # Synchronous
        result = await adapter.execute(
            ExecuteRequest(message="What is 2+2?")
        )
        print(result.content)
        ```
    """

    def __init__(
        self,
        api_key: str | None = None,
        model: str = "claude-sonnet-4-20250514",
        max_tokens: int = 4096,
        tools: list[ToolDefinition] | None = None,
        tool_handlers: dict[str, Any] | None = None,
    ):
        """
        Initialize the Anthropic adapter.

        Args:
            api_key: Anthropic API key (defaults to ANTHROPIC_API_KEY env var)
            model: Model to use (default: claude-sonnet-4-20250514)
            max_tokens: Maximum tokens in response
            tools: Tool definitions for the model
            tool_handlers: Functions to handle tool calls {name: callable}
        """
        self._api_key = api_key or os.environ.get("ANTHROPIC_API_KEY")
        if not self._api_key:
            raise ValueError(
                "Anthropic API key required. Set ANTHROPIC_API_KEY or pass api_key."
            )

        self._model = model
        self._max_tokens = max_tokens
        self._tools = tools or []
        self._tool_handlers = tool_handlers or {}
        self._client = anthropic.AsyncAnthropic(api_key=self._api_key)
        self._messages: list[dict[str, Any]] = []

    @property
    def id(self) -> str:
        return "anthropic-agent"

    @property
    def name(self) -> str:
        return "Anthropic Agent SDK"

    @property
    def version(self) -> str:
        return "0.1.0"

    @property
    def capabilities(self) -> AdapterCapabilities:
        return AdapterCapabilities(
            agents=False,
            execution=True,
            streaming=True,
            sessions=True,  # In-memory conversation
            memory=False,
            subagents=False,
            mcp=False,
            files=False,
            hooks=False,
            planning=False,
            skills=False,
            websocket=False,
            multipart=False,
            binary_download=False,
        )

    async def get_capability_manifest(self) -> CapabilityManifest:
        return CapabilityManifest(
            harness_id=self.id,
            version=self.version,
            capabilities=[
                CapabilityInfo(id="execution.run", supported=True),
                CapabilityInfo(id="execution.stream", supported=True),
                CapabilityInfo(id="execution.thinking", supported=True),
                CapabilityInfo(id="tools.register", supported=True),
                CapabilityInfo(id="tools.invoke", supported=True),
                CapabilityInfo(id="sessions.create", supported=True),
                CapabilityInfo(id="sessions.history", supported=True),
            ],
        )

    def _format_tools(self) -> list[dict[str, Any]]:
        """Format tools for Anthropic API."""
        return [
            {
                "name": t.name,
                "description": t.description,
                "input_schema": t.input_schema,
            }
            for t in self._tools
        ]

    async def execute(
        self,
        request: ExecuteRequest,
        **options: Any,
    ) -> AdapterExecutionResult:
        """Execute a prompt and return the complete result."""
        # Add user message
        self._messages.append({"role": "user", "content": request.message})

        try:
            # Build request
            kwargs: dict[str, Any] = {
                "model": self._model,
                "max_tokens": self._max_tokens,
                "messages": self._messages,
            }

            if request.system_prompt:
                kwargs["system"] = request.system_prompt

            if self._tools:
                kwargs["tools"] = self._format_tools()

            # Make request
            response = await self._client.messages.create(**kwargs)

            # Process response
            output_parts = []
            tool_calls = []

            for block in response.content:
                if block.type == "text":
                    output_parts.append(block.text)
                elif block.type == "tool_use":
                    tool_calls.append({
                        "id": block.id,
                        "name": block.name,
                        "input": block.input,
                    })

            # Add assistant message to history
            self._messages.append({"role": "assistant", "content": response.content})

            # Handle tool calls if we have handlers
            if tool_calls and self._tool_handlers:
                tool_results = []
                for tc in tool_calls:
                    handler = self._tool_handlers.get(tc["name"])
                    if handler:
                        result = await handler(**tc["input"]) if callable(handler) else str(handler)
                        tool_results.append({
                            "type": "tool_result",
                            "tool_use_id": tc["id"],
                            "content": str(result),
                        })

                if tool_results:
                    self._messages.append({"role": "user", "content": tool_results})
                    # Recurse to get final response
                    return await self.execute(ExecuteRequest(message=""), **options)

            usage = None
            if response.usage:
                usage = UsageStats(
                    input_tokens=response.usage.input_tokens,
                    output_tokens=response.usage.output_tokens,
                    total_tokens=response.usage.input_tokens + response.usage.output_tokens,
                )

            return AdapterExecutionResult(
                output="\n".join(output_parts),
                tool_calls=tool_calls,
                usage=usage,
                metadata={"model": response.model, "stop_reason": response.stop_reason},
            )

        except Exception as e:
            return AdapterExecutionResult(
                output="",
                tool_calls=[],
                usage=None,
                metadata={"error": str(e)},
            )

    async def execute_stream(
        self,
        request: ExecuteRequest,
        **options: Any,
    ) -> AsyncIterator[ExecutionEvent]:
        """Execute with streaming."""
        # Add user message
        self._messages.append({"role": "user", "content": request.message})

        try:
            # Build request
            kwargs: dict[str, Any] = {
                "model": self._model,
                "max_tokens": self._max_tokens,
                "messages": self._messages,
            }

            if request.system_prompt:
                kwargs["system"] = request.system_prompt

            if self._tools:
                kwargs["tools"] = self._format_tools()

            # Stream response
            current_tool: dict[str, Any] | None = None
            response_content: list[Any] = []

            async with self._client.messages.stream(**kwargs) as stream:
                async for event in stream:
                    if event.type == "content_block_start":
                        block = event.content_block
                        if block.type == "text":
                            pass  # Text will come in deltas
                        elif block.type == "tool_use":
                            current_tool = {
                                "id": block.id,
                                "name": block.name,
                                "input": {},
                            }
                            yield ToolCallStartEvent(
                                id=block.id,
                                name=block.name,
                                input={},
                            )
                        elif block.type == "thinking":
                            pass  # Thinking will come in deltas

                    elif event.type == "content_block_delta":
                        delta = event.delta
                        if delta.type == "text_delta":
                            yield TextEvent(content=delta.text)
                        elif delta.type == "thinking_delta":
                            yield ThinkingEvent(thinking=delta.thinking)
                        elif delta.type == "input_json_delta":
                            if current_tool:
                                # Accumulate JSON input
                                pass

                    elif event.type == "content_block_stop":
                        if current_tool:
                            yield ToolCallEndEvent(id=current_tool["id"])
                            response_content.append(current_tool)
                            current_tool = None

                # Get final message for history
                final_message = await stream.get_final_message()
                self._messages.append({"role": "assistant", "content": final_message.content})

                usage = None
                if final_message.usage:
                    usage = UsageStats(
                        input_tokens=final_message.usage.input_tokens,
                        output_tokens=final_message.usage.output_tokens,
                        total_tokens=final_message.usage.input_tokens + final_message.usage.output_tokens,
                    )

                yield DoneEvent(usage=usage)

        except Exception as e:
            yield ErrorEvent(
                code="api_error",
                message=str(e),
                recoverable=False,
            )

    # Session management

    def clear_history(self) -> None:
        """Clear conversation history."""
        self._messages.clear()

    def get_history(self) -> list[dict[str, Any]]:
        """Get conversation history."""
        return self._messages.copy()

    # Tool management

    async def list_tools(self) -> list[Tool]:
        """List registered tools."""
        return [
            Tool(
                id=t.name,
                name=t.name,
                description=t.description,
                input_schema=t.input_schema,
            )
            for t in self._tools
        ]

    async def register_tool(self, tool: ToolDefinition) -> None:
        """Register a tool."""
        self._tools.append(tool)

    async def unregister_tool(self, tool_id: str) -> None:
        """Unregister a tool."""
        self._tools = [t for t in self._tools if t.name != tool_id]

    def set_tool_handler(self, name: str, handler: Any) -> None:
        """Set a handler function for a tool."""
        self._tool_handlers[name] = handler

    # Cleanup

    async def close(self) -> None:
        """Clean up resources."""
        await self._client.close()
        self._messages.clear()
